### DISCUSSION TITLE
Ideas for resurrecting the plugin

### DISCUSSION THREAD

**Discussion by P403n1x87** [P403n1x87/pytest-austin#5] [Ideas]
The latest release of this pytest plugin for Austin included support for Austin 2. These are a few ideas to add support for the latest Austin release and to improve the way the plugin works in general.

## Implementation

The current approach is to spawn Austin and attach to the running pytest process. This requires that the test suite is run with ptrace capabilities. Furthermore this requires the extra effort of extracting the sub-flamegraph from the full dataset. A few alternative approaches are based on the idea of executing marked tests in a subprocess:

- [pytest-forked](https://github.com/pytest-dev/pytest-forked) is the standard approach, but lacks support for platforms with no `fork` (e.g. Windows).
- The [dd-trace-py](https://github.com/P403n1x87/dd-trace-py/blob/ac30ec213b9f6536106858ff67e6a0282f6cb393/conftest.py#L31-L48) test suite has an AST-based approach for extracting the code from a test, and compiling it into bytecode to run in a subprocess

The second approach comes with a few challenges to overcome some limitations, but makes for a cross-platform solution and therefore is the preferred one.

## Support

- Austin 3.5 has quite a few differences from 2.0, most notably the command line interface and a more efficient binary output mode.

**Comment by P403n1x87** [comment]
@jayaddison I've started this discussion to collect a few ideas on how to resurrect this plugin.

**Reply by jayaddison** [reply]
Looks good - thanks @P403n1x87.  The first question, I suppose, is what version(s) of Austin to support during this refresh work.

Also: could you fixup the link for `dd-trace-py`, for future readers?

**Reply by P403n1x87** [reply]
> Looks good - thanks @P403n1x87. The first question, I suppose, is what version(s) of Austin to support during this refresh work.

I think it's fine to support the latest, i.e. 3.5.

> Also: could you fixup the link for `dd-trace-py`, for future readers?

Ah my bad, I didn't realise that link got butchered. Should be fixed now.

**Reply by jayaddison** [reply]
Ok, thanks for correcting that.

Do you have a sense for whether it would be significantly more work to support multiple versions of Austin from each major version of `pytest-austin`?  The reason I ask is that, design-wise, we may also want to decide whether to tightly-couple the functionality here with Austin release versions -- or we could decide to support version ranges.

Initially I feel that it may be simpler to match major versions between Austin and `pytest-austin`; but I'm not too familiar with the interface between the two components yet (command-line?  lower-level Python interfaces, as suggested by the AST option?).

**Reply by P403n1x87** [reply]
My opinion on this is that, In the ideal world, any new minor/major release of Austin should be accompanied by new releases of all the libraries and tools in the Austin-sphere (e.g. the TUI, the Python wrapper library, this plugin etc...). However, this is unlikely to happen because I don't have the capacity to work on all these projects in a timely manner. As this plugin shows, work on these tools is prioritised based on my sense of where the higher benefit is. The traffic on this repository has been very low, which makes me think that this plugin is not used a lot, and that's why I didn't receive much attention from me lately.

I believe the best course of action is to support just the latest major release of Austin. With the guarantees of semantic versioning, we can expect each release of pytest-austin to always support the latest minor releases of Austin (the only limitation would be that new features/improvements won't be available). A new major release of Austin won't necessarily require a major release of pytest-austin, unless the latter also brings some breaking changes in its API.

**Reply by jayaddison** [reply]
A brief update - not much progress so far: I've cloned the `dd-trace-py` source locally, and intend to learn how the AST-based approach mentioned is implemented in that library.

**Reply by P403n1x87** [reply]
Hey, absolutely no rush on this! 🙂 . You probably have found this already, but just in case the implementation of the pytest marker is in these lines
https://github.com/P403n1x87/dd-trace-py/blob/b29bad5141d0f9d5145ef5a778e98ec7f441754e/tests/conftest.py#L120-L297
The AST "capturing" is essentially done by the `FunctionDefFinder` class.

**Reply by jayaddison** [reply]
Well, approximately `P2M` (ISO-8601 duration) later.. I think I'll be ready to take another look at this within the next few days.

**Reply by jayaddison** [reply]
(and I notice that v3.6.0 of `austin` has been released, so that'll be the version I'll aim to develop against for `pytest-austin`)

**Reply by P403n1x87** [reply]
There are no big differences between 3.5 and 3.6, apart a bump in the MOJO version (now using version 3), and the drop of support for legacy Python version, which should actually make this a bit less of a pain since there's less to support from the plugin side 🙂

**Reply by jayaddison** [reply]
Before getting onto the AST-parsing idea: I've found that the reason that the plugin currently stalls (`austin` is opened as a subprocess, but never terminated) is due to the use of [v0.1.0 of `austin-python` as a dependency](https://github.com/P403n1x87/pytest-austin/blob/3c8abcfa3e7c3b9520a82effb4df86fc1a44b137/pyproject.toml#L53).

This causes the [`_read_header` method to return `False`](https://github.com/P403n1x87/austin-python/blob/5759ea15651669ad22b8f6efe7b6dd9cd6677da7/austin/simple.py#L60-L69) because it doesn't match against the updated output of `austin` v3.6.0 (or v3.5.0 that I was using previously).

**Reply by jayaddison** [reply]
...OK, and ~~that's~~ after fixing that, there's another problem during output parsing: because the line format emitted by the `austin` binary -- the format being MOJO -- has changed.  It looks like the plugin previously expected a single-metric-per-line, however `austin` can now emit one-or-more metrics per line.

**Reply by P403n1x87** [reply]
Ah that's right. But all the details about the MOJO encoding should be well encapsulated in the latest version of the austin-python version https://github.com/P403n1x87/austin-python/pull/18 🤞

**Reply by P403n1x87** [reply]
Spec here https://github.com/P403n1x87/austin/wiki/The-MOJO-File-Format

**Reply by P403n1x87** [reply]
I had some time to formalise the AST idea into a draft PR https://github.com/P403n1x87/pytest-austin/pull/6. I haven't tested this yet, but currently there isn't much to test as the main logic for processing and checking the collected data still needs to be ported over.

**Reply by jayaddison** [reply]
Thanks for including me in the update!  I've taken a look at the changes to try to learn a bit more.

If I understand: currently the changes run `austin` single-threaded, I guess so that it's simpler to test and get the MOJO conversion code correct, before reinstating threading.

And I see the the `ast` parsing is used to read the contents of `pytest-austin`-markeded function and compile that to file that is then passed to the `austin` binary as an argument.

Something I'm wondering about: how are dependencies/imports handled when `compile` is used on a subset of a file?  (I'm comparing the `dd-trace-py` code to see how `FunctionDefFinder` is used there, but I don't see where it's referenced yet.  I should experiment with it locally to see how it behaves)

**Reply by jayaddison** [reply]
Also: I _think_ [this is where](https://github.com/P403n1x87/austin-python/blob/44f1c4af86b2f462b5005f57311b19d6a635e44b/austin/format/mojo.py#L5) some of the utility methods to do the MOJO conversion should be sourced from?

(my approximate understanding so far is that it's a terse line-based format for encoding trace metrics -- but not familiar with the details of it yet.  I did scan the documentation, but haven't grokked it yet)

**Reply by P403n1x87** [reply]
> If I understand: currently the changes run austin single-threaded, I guess so that it's simpler to test and get the MOJO conversion code correct, before reinstating threading.

That's right. We use austin to span a subprocess with the test logic synchronously. This means that we don't have to run a separate thread, nor have Austin attach to a running process, which also resolves the permission issues on Linux (that is, we no longer need to run with sudo or set capabilities etc...).

> And I see the the ast parsing is used to read the contents of pytest-austin-markeded function and compile that to file that is then passed to the austin binary as an argument.

Correct!

> Something I'm wondering about: how are dependencies/imports handled when compile is used on a subset of a file?

The idea of the marker is to turn tests that would run a Python script (in the form of an embedded literal string) in a subprocess to something that looks like actual Python code in an ordinary test script. This means that imports will have to be included in the test function itself. E.g.

```python
@pytest.mark.austin(...)
def test_frame_cpu_constraint():
    import mylibrary.foo

    foo()
```

> Also: I think [this is where](https://github.com/P403n1x87/austin-python/blob/44f1c4af86b2f462b5005f57311b19d6a635e44b/austin/format/mojo.py#L5) some of the utility methods to do the MOJO conversion should be sourced from?

That's right. The logic for handling a MOJO file is already implemented in austin-python.

> (my approximate understanding so far is that it's a terse line-based format for encoding trace metrics -- but not familiar with the details of it yet. I did scan the documentation, but haven't grokked it yet)

Yep. It's a binary format for encoding raw Austin events as they occur with the lowest possible overhead and with the simplest redundancy removal baked in. Hopefully the interface exposed by the `MojoFile` reader in austin-python makes the knowledge of the format spec not necessary, as the data should be available in the form of stacks and metadata from a `MojoFile` instance.

**Reply by jayaddison** [reply]
Ok!  Thank you @P403n1x87.  As usual, I'm going to be a bit unreliable here and likely not respond again for a few weeks.  I am still following this though (and hoping to help out if I can).
